music21/musicxml/m21ToXml.py

```diff
@@ -3038,25 +3038,46 @@ def parseFlatElements(self, m, *, backupAfterwards=False):
 
         self.currentVoiceId = voiceId
 
-        # group all objects by offsets and then do a different order than normal sort.
-        # that way chord symbols and other 0-width objects appear before notes as much as
-        # possible.
-        for objGroup in OffsetIterator(m):
-            groupOffset = m.elementOffset(objGroup[0])
+        # Group just notes and rests (GeneralNote) by their offsets so that
+        # <forward> tags (which become hidden rests) only bridge gaps between
+        # notes and rests, and not gaps formed with expressions that can
+        # be represented instead with <offset> tags.
+        # See m. 1 of Weber piece in core corpus
+        general_note_offsets = []
+        for objGroup in OffsetIterator(m, filterList=[lambda n: isinstance(n, note.GeneralNote)]):
+            general_note_offsets.append(m.elementOffset(objGroup[0]))
+        if not general_note_offsets:
+            # fallback if there are no notes or rests: this will capture DaCapo signs, etc
+            general_note_offsets = [0.0]
+        for i, groupOffset in enumerate(general_note_offsets):
             amountToMoveForward = int(round(divisions * (groupOffset
                                                              - self.offsetInMeasure)))
             if amountToMoveForward > 0:
-                # gap in stream!
+                # gap in stream between GeneralNote objects: create <forward>
                 mxForward = Element('forward')
                 mxDuration = SubElement(mxForward, 'duration')
                 mxDuration.text = str(amountToMoveForward)
                 root.append(mxForward)
                 self.offsetInMeasure = groupOffset
 
             notesForLater = []
-            for obj in objGroup:
+            is_last_gn_offset = i + 1 == len(general_note_offsets)
+            if is_last_gn_offset:
+                nextGroupOffset = m.highestTime
+            else:
+                nextGroupOffset = general_note_offsets[i + 1]
+            for obj in m.getElementsByOffset(
+                offsetStart=groupOffset,
+                offsetEnd=nextGroupOffset,
+                includeEndBoundary=is_last_gn_offset,
+                mustFinishInSpan=False,
+                mustBeginInSpan=True,
+                includeElementsThatEndAtStart=True,
+            ):
```

Comment:
getElementsByOffset is pretty slow -- and it ends up as a for-loop within a for-loop for what most of the time would be a no-op.

I wonder if there's a way to use the older code, getting everything in the OffsetIterator but only emitting a `<forward>` tag and setting offsetInMeasure if `any(isinstance(n, GeneralNote) for n in offsetGroup)`.  And making a separate list of offsetGroups with notes and without (or better, iterate over a cheap dataclass of `.offset`, `.elements`, `.hasSomeNotes`, `.isLast`, `.distanceToNextGroup`)  -- Does this make sense?

Reply:
er... `.distanceToNextGroupHavingNotes` I suppose.

Reply:
Smart, and it looks like just doing the `isinstance` check is enough.